kubernetes-client/src/main/java/io/fabric8/kubernetes/client/URLFromNodePortImpl.java

```diff
@@ -0,0 +1,68 @@
+/**
+ * Copyright (C) 2015 Red Hat, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.fabric8.kubernetes.client;
+
+import io.fabric8.kubernetes.api.model.*;
+import io.fabric8.kubernetes.client.utils.URLFromServiceImplUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.List;
+
+public class URLFromNodePortImpl implements ServiceToURLProvider {
+  public static Logger logger = LoggerFactory.getLogger(URLFromNodePortImpl.class);
+
+  public String getURL(Service service, String portName, String namespace, KubernetesClient client) {
+    boolean bFound = false;
+    ServicePort port = URLFromServiceImplUtil.getServicePortByName(service, portName);
+    String serviceProto = port.getProtocol();
+    String clusterIP = null;
+    Integer portNumber = 0;
+    Integer nodePort = port.getNodePort();
+    if(nodePort != null) {
+      try {
+        NodeList nodeList = client.nodes().list();
+        if(nodeList != null && nodeList.getItems() != null) {
+          for(Node item : nodeList.getItems()) {
+            NodeStatus status = item.getStatus();
+            if(!bFound && status != null) {
+              List<NodeAddress> addresses = status.getAddresses();
+              for(NodeAddress address : addresses) {
```

Comment:
Can you put this second for inside a method? Just do not create two nested `for` directly. I usually try to avoid this since for a reader is easier to check what this method exactly do. Since Java does not have multi return like go, this new method could return a simple private class with required arguments.

Reply:
ah, I wish we had already upgraded this project to Java 8, I could have used ` javafx.util.Pair ` :smile: